File: leetcode/python/925. 长按键入.py
#!/bin/bash python
"""
925. 长按键入

思路：双指针法
"""

class Solution(object):
    def isLongPressedName(self, name, typed):
        """
        双指针
        """
        p = q = 0
        while p < len(name) and q < len(typed):
            if name[p] == typed[q]:
                q += 1
                p = min(p + 1, len(name) - 1)
            elif p - 1 >= 0 and name[p - 1] == typed[q]:
                q += 1
            else:
                return False
        return p == len(name) - 1 and q == len(typed) and name[p] == typed[q - 1]

if __name__ == "__main__":
    name = "pyplrz"
    typed = "ppyypllr"
    s = Solution()
    print(s.isLongPressedName(name, typed))

File: leetcode/cpp/925. 长按键入.cpp
#include <stdio.h>
#include <string>

class Solution {
public:
    bool isLongPressedName(std::string name, std::string typed) {
        int p = 0;
        int q = 0;
        while (p < name.size() && q < typed.size()) {
            if (name[p] == typed[q]) {
                q += 1;
                p += 1;
                if (p > name.size() - 1) {
                    p = name.size() - 1;
                }
            } else if (p - 1 >= 0 && name[p - 1] == typed[q]) {
                q += 1;
            } else {
                return false;
            }
        }
        return p == name.size() - 1 && q == typed.size() && name[p] == typed[q - 1];
    }
};

int main(int argc, char* argv[])
{
    std::string name = "pyplrz";
    std::string typed = "ppyypllr";
    Solution s;
    printf("%d\n", s.isLongPressedName(name, typed));
    return 0;
}